frontend/test/pytest/test_jax_dynamic_api.py

```diff
@@ -0,0 +1,385 @@
+from numpy import array_equal
+from numpy.testing import assert_allclose
+
+from catalyst import qjit, while_loop
+
+DTYPES = [float, int, jnp.float32, jnp.float64, jnp.int8, jnp.int16, "float32", np.float64]
+SHAPES = [3, (2, 3, 1), (), jnp.array([2, 1, 3], dtype=int)]
+
+
+def assert_array_and_dtype_equal(a, b):
+    """Check that two arrays have exactly the same values and types"""
+
+    assert array_equal(a, b)
+    assert a.dtype == b.dtype
+
+
+def test_qjit_abstracted_axes():
+    """Test that qjit accepts dynamical arguments."""
+
+    @qjit(abstracted_axes={0: "n"})
+    def identity(a):
+        return a
+
+    param = jnp.array([1, 2, 3])
+    result = identity(param)
+    assert_array_and_dtype_equal(param, result)
+    assert "tensor<?xi64>" in identity.mlir, identity.mlir
+
+
+def test_qnode_abstracted_axis():
+    """Test that qnode accepts dynamical arguments."""
+
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def circuit(a):
+        return a
+
+    @qjit(abstracted_axes={0: "n"})
+    def identity(a):
+        return circuit(a)
+
+    param = jnp.array([1, 2, 3])
+    result = identity(param)
+
+    assert_array_and_dtype_equal(param, result)
+    assert "tensor<?xi64>" in identity.mlir, identity.mlir
+
+
+def test_qnode_dynamic_structured_args():
+    """Test that qnode accepts dynamically-shaped structured args"""
+
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def circuit(a_b):
+        return a_b[0] + a_b[1]
+
+    @qjit(abstracted_axes={0: "n"})
+    def func(a_b):
+        return circuit(a_b)
+
+    param = jnp.array([1, 2, 3])
+    c = func((param, param))
+    assert_array_and_dtype_equal(c, param + param)
+    assert "tensor<?xi64>" in func.mlir, func.mlir
+
+
+def test_qnode_dynamic_structured_results():
+    """Test that qnode returns dynamically-shaped results"""
+
+    @qml.qnode(qml.device("lightning.qubit", wires=1))
+    def circuit(a):
+        return (
+            jnp.ones((a + 1,)),
+            jnp.ones(
+                (a + 2),
+            ),
+        )
+
+    @qjit
+    def func(a):
+        return circuit(a)
+
+    a, b = func(3)
+    assert_allclose(a, jnp.ones((4,)))
+    assert_allclose(b, jnp.ones((5,)))
+    assert "tensor<?xf64>" in func.mlir, func.mlir
+
+
+@pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("shape", SHAPES)
+def test_classical_tracing_init(shape, dtype):
+    """Test that tensor primitive work in the classical tracing mode"""
+
+    assert_array_and_dtype_equal(
+        qjit(lambda: jnp.zeros(shape, dtype))(), jnp.zeros(shape, dtype=dtype)
+    )
+    assert_array_and_dtype_equal(
+        qjit(lambda: jnp.ones(shape, dtype))(), jnp.ones(shape, dtype=dtype)
+    )
+    assert_array_and_dtype_equal(
+        qjit(lambda s: jnp.ones(s, dtype))(shape), jnp.ones(shape, dtype=dtype)
+    )
+    assert_array_and_dtype_equal(
+        qjit(lambda s: jnp.zeros(s, dtype))(shape), jnp.zeros(shape, dtype=dtype)
+    )
```

Comment:
Weren't these going to be removed? (I remember a previous comment saying they were from early testing)

Reply:
No, the previous ones were without the qjit wrappers. The idea was to test a regular Python interpretation against the custom primitives we had at the moment. These tests do test basic jnp.zeros/ones functions in classical tracing mode

Reply:
Ah good point, my bad 👍